kernel/mempool.c

```diff
@@ -53,7 +53,20 @@ void k_free(void *ptr)
 
 		k_heap_free(*heap_ref, ptr);
 
+		/* As you can see above, we reference `heap_ref` from `ptr`, which is fine on its
+		 * own. However, the GCC malloc attribute does not allow any aliasing of pointers,
+		 * causing GCC to issue a warning when `heap_ref` is used in the tracing macro after
+		 * calling `k_heap_free(*heap_ref, ptr)`.
+		 *
+		 * The warning can be suppressed by using -Wuse-after-free, but unfortunately, this
+		 * option is only available for GCC >= 12.
+		 */
+		#pragma GCC diagnostic push
+		#if defined __GNUC__ && __GNUC__ >= 12
+		#pragma GCC diagnostic ignored "-Wuse-after-free"
+		#endif
 		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_heap_sys, k_free, *heap_ref, heap_ref);
+		#pragma GCC diagnostic pop
```

Comment:
Is this a good fix?

Reply:
I think this should be declared in the toolchain files. Take a look at `TOOLCHAIN_IGNORE_WSHADOW` and mimic that?

In `include/zephyr/toolchain/gcc.h`:

```c
#if defined __GNUC__ >= 12
#define TOOLCHAIN_IGNORE_WUSE_AFTER_FREE_BEGIN \
	_Pragma("GCC diagnostic push") \
	_Pragma("GCC diagnostic ignored \"-Wuse-after-free\"")

#define TOOLCHAIN_IGNORE_WUSE_AFTER_FREE_END \
	_Pragma("GCC diagnostic pop")
#endif
```